### ISSUE TITLE
Unclean exit from ipython on MacOS

### ISSUE THREAD

**Issue by mlxd** [microsoft/iqsharp#39]
**Describe the bug**

Hi, I have installed the qsharp package as described here: https://docs.microsoft.com/en-us/quantum/install-guide/python?view=qsharp-preview and have successfully run the Python examples given on my machine in an IPython environment. Hoewever, on exit I receive the following traceback:

```python
Exception ignored in: <function Socket.__del__ at 0x1041b4290>
Traceback (most recent call last):
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 67, in __del__
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 105, in close
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/context.py", line 153, in _rm_socket
TypeError: 'NoneType' object is not callable
Exception ignored in: <function Socket.__del__ at 0x1041b4290>
Traceback (most recent call last):
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 67, in __del__
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 105, in close
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/context.py", line 153, in _rm_socket
TypeError: 'NoneType' object is not callable
Exception ignored in: <function Socket.__del__ at 0x1041b4290>
Traceback (most recent call last):
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 67, in __del__
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 105, in close
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/context.py", line 153, in _rm_socket
TypeError: 'NoneType' object is not callable
```

**To Reproduce**
To attempt to rule out my python environment, I set up a clean miniconda build and ran the example from there:
```bash
cd ~ && mkdir qstmp && cd qstmp
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
chmod +x ./Miniconda3-latest-MacOSX-x86_64.sh
./Miniconda3-latest-MacOSX-x86_64.sh -b -p ./qs_conda
source ./qs_conda/bin/activate ;
conda create -n qs_env -y python
conda activate qs_env
dotnet tool install -g Microsoft.Quantum.IQSharp
dotnet iqsharp install
dotnet iqsharp --version
dotnet tool update -g Microsoft.Quantum.IQSharp
pip install qsharp --upgrade
conda install ipython -y
ipython -c "import qsharp"
```

The result was as listed above.  Running `ipython -c "import zmq"` gave no output.

**Expected behavior**
There should be no output/traceback from importing the qsharp module and exiting.

**System information**
 - OS: MacOS 10.14.6
 - iqsharp: 0.9.1908.2906
 - Jupyter Core: 1.1.14623.0

If I can give any more input, feel free to let me know.

Thanks,
Lee.

**Comment by bromeg** [comment]
Thank you for reporting this! We're having some trouble reproducing this reliably on our end, but we are actively investigating what might be causing this bug.

**Comment by anpaz-msft** [comment]
I've started to see this too. It's unclear of what the problem is. Seems to be something in the `zmq` library that ipython uses for communication with kernels but not sure what.
This seems to happen only on Macs and only during exit.
Getting help on figuring out what is causing this error is greatly appreciated!

**Comment by mlxd** [comment]
I had a look around for some info and added logging info to the methods in "zmq/sugar/socket.py" (https://github.com/zeromq/pyzmq/blob/master/zmq/sugar/socket.py, attached as 
[socket.txt](https://github.com/microsoft/iqsharp/files/3775081/socket.txt) )

The output may be showing that one socket is attempting to be closed after having already been closed previously (though I could be mistaken). Output from using the above `socket.py` when running `ipython -c "import qsharp"` for me now gives:

```bash
Creating socket with params: a=(<zmq.sugar.context.Context object at 0x10715e050>, 5) kw=dict_items([])
Creating socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 2) kw=dict_items([])
Creating socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 5) kw=dict_items([])
Creating socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 5) kw=dict_items([])
Closing socket with params: a=(<zmq.sugar.context.Context object at 0x10715e050>, 5) kw=dict_items([]) linger=None already_closed=False
Deleting socket with params: a=(<zmq.sugar.context.Context object at 0x10715e050>, 5) kw=dict_items([])
Closing socket with params: a=(<zmq.sugar.context.Context object at 0x10715e050>, 5) kw=dict_items([]) linger=None already_closed=True
Deleting socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 2) kw=dict_items([])
Closing socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 2) kw=dict_items([]) linger=None already_closed=False
Exception ignored in: <function Socket.__del__ at 0x106bd00e0>
Traceback (most recent call last):
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 73, in __del__
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 116, in close
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/context.py", line 153, in _rm_socket
TypeError: 'NoneType' object is not callable
Deleting socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 5) kw=dict_items([])
Closing socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 5) kw=dict_items([]) linger=None already_closed=False
Exception ignored in: <function Socket.__del__ at 0x106bd00e0>
Traceback (most recent call last):
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 73, in __del__
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 116, in close
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/context.py", line 153, in _rm_socket
TypeError: 'NoneType' object is not callable
Deleting socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 5) kw=dict_items([])
Closing socket with params: a=(<zmq.sugar.context.Context object at 0x107135e30>, 5) kw=dict_items([]) linger=None already_closed=False
Exception ignored in: <function Socket.__del__ at 0x106bd00e0>
Traceback (most recent call last):
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 73, in __del__
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/socket.py", line 116, in close
  File "/Users/mlxd/qstmp/qs_conda/envs/qs_env/lib/python3.7/site-packages/zmq/sugar/context.py", line 153, in _rm_socket
TypeError: 'NoneType' object is not callable
```

I am not too familiar with zmq or how it integrates with the project, but I can continue looking around if that helps.

**Comment by numpde** [comment]
I'm on Ubuntu 18.04 / Python 3.7.5, have the same symptoms.

**Comment by cqc-alec** [comment]
I confirm this is not a Mac-only issue. Ubuntu 18.04, ipython 7.8.0, same issue with `ipython -c "import qsharp"`.

**Comment by cgranade** [comment]
As per a report from @dbwz8, the same error happens on Windows, at least in some configurations. We're still trying to nail down a minimum working reproduction in order to solve this issue. Thank you everyone for your patience!

**Comment by rmshaffer** [comment]
This appears to be a garbage collection issue in the pyzmq project, tracked by https://github.com/zeromq/pyzmq/issues/1356. It is harmless, since it's just a failure during garbage collection of something that had already been released.

According to the PR https://github.com/zeromq/pyzmq/pull/1367, it is fixed in version 19.0.1 of pyzmq. I verified this by running `pip install pyzmq==19.0.1` and noting that these errors on shutdown stopped occurring.

Closing this issue since it's external, benign, and is already fixed in the upstream package.
